packages/fast-components/src/design-system-provider/design-system-provider.ts

```diff
@@ -1,98 +1,271 @@
-import { parseColorHexRGB } from "@microsoft/fast-colors";
-import {
-    createColorPalette,
-    DesignSystem,
-    DesignSystemDefaults,
-} from "@microsoft/fast-components-styles-msft";
-import { FastElement, observable, Observable } from "@microsoft/fast-element";
+import { FastElement, Observable } from "@microsoft/fast-element";
+import { DesignSystemConsumer, designSystemConsumer } from "../design-system-consumer";
+
+interface CustomPropertyDefinition {
+    name: string;
+    value: string | (() => string);
+}
+
+interface DesignSystemPropertyDeclarationConfig {
+    customPropertyName?: string;
+    customProperty?: boolean;
+}
 
 /**
- * The following interface is an implementation of
- * design-system.schema.json in TypeScript with
- * as much accuracy as TypeScript supports.
+ * Decorator to declare a property as a design-system property.
+ * Accepts an optional config to customize whether a css custom property
+ * wll be written and if so, what the name of that property is.
  */
-export interface DesignSystemData {
-    accentBaseColor: string;
-    backgroundColor: string;
-    cornerRadius: number;
-    disabledOpacity: number;
-    density: number;
-    designUnit: number;
-    elevatedCornerRadius: number;
-    focusOutlineWidth: number;
-    outlineWidth: number;
+export function designSystemProperty<T extends DesignSystemProvider>(
+    config: DesignSystemPropertyDeclarationConfig
+): (source: T, property: string) => void;
+export function designSystemProperty<T extends DesignSystemProvider>(
+    source: T,
+    property: string
+): void;
+export function designSystemProperty<T extends DesignSystemProvider>(

```

Comment:
I find this style of functions really hard to read and understand. Is there no better / new way to achieve this, or should we create named variants instead or something? The logic is often hard to follow.

Reply:
You mean the overload syntax? https://www.typescriptlang.org/docs/handbook/functions.html#overloads

There isn't really a better alternative that I know of.